amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -191,26 +191,31 @@ final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider,
         state.compareAndSet(Connecting, Connected(connection, 1))
         connection
       } else get
+    case Connecting => get
     case c @ Connected(connection, clients) =>
       if (state.compareAndSet(c, Connected(connection, clients + 1))) connection
       else get
-    case _ => get
+    case Closing => get
+    case _ => throw new RuntimeException("Unknown provider state.")
   }
 
   @tailrec
   override def release(connection: Connection): Unit = state.get match {
+    case Empty => ()
```

Comment:
Isn't this an illegal-state one as well, how could it be `Empty` when we release unless something is wrong?

Reply:
If the user call release before the connection is even opened.

We could:
* Throw an exception stating that you can't release if the connection hasn't been opened.
* Just let it be. It's already released so there is nothing to do

I chose option B. Would you prefer option A?

Reply:
Yes, absolutely option A, because it would either be a bug in our concurrency logic and be _us_ calling it or somehow a bug in the user logic, either way we'd not want it to silently behave as everything is ok but to detect that something was not what we thought it is.